contracts/ERC1155M.sol

```diff
@@ -523,6 +568,52 @@ contract ERC1155M is IERC1155M, ERC1155Supply, ERC2981, Ownable2Step, Reentrancy
         revert InvalidStage();
     }
 
+     /**
+     * @dev Returns data hash for the given minter, qty and timestamp.
+     */
+    function getCosignDigest(
+        address minter,
+        uint256 tokenId,
+        uint32 qty,
+        uint64 timestamp
+    ) public view returns (bytes32) {
+        if (_cosigner == address(0)) revert CosignerNotSet();
+
+        return
+            MessageHashUtils.toEthSignedMessageHash(
+                keccak256(
+                    abi.encodePacked(
+                        address(this),
+                        minter,
+                        qty,
+                        _cosigner,
+                        timestamp,
+                        block.chainid,
+                        getCosignNonce(minter, tokenId)
```

Comment:
I think there may be a potential transaction ordering issue here. @channing-magiceden Is there any kind of ordering protection here?

Would the offchain signature need to calculate all previous mints, including unredeemed/pending transactions, to have the correct nonce value? 
Can a mint from `authorizedMint` with `to` set to a user address break the nonce in an existing signature? E.g. a race condition?
What happens if a user abandons a tx that has been cosigned?
Do all successive mints depend on transaction ordering? So if a tx with a low fee does not mine, do all successive txs fail because the nonce is incorrect?

Reply:
Good call.

> Would the offchain signature need to calculate all previous mints, including unredeemed/pending transactions, to have the correct nonce value?
Yes offchain signature needs to count all preivous mints.

> Can a mint from authorizedMint with to set to a user address break the nonce in an existing signature? E.g. a race condition?
Yes, minting activity will bump nonce so it invalidates the old signature. This is the point of cosigning - we want to rate limit and prevent bots by controlling the cosigning process. Our UI / API will fetch new signature in that case. e.g. https://github.com/magiceden/magiceden/blob/84a82917a97cd6924a0c2fcbeb4cc42a3d73342d/js/packages/api/src/v3/routes/frame.route.ts#L67-L78

> What happens if a user abandons a tx that has been cosigned?
If mint fails or not happen, the nonce won't change. User can reuse the signature (if not expired), or grab new signature from cosign service.

> Do all successive mints depend on transaction ordering?
I don't think so. This nonce is not bond to tx but to `totalMintedByAddress`. If previous tx fails, the nonce won't bump.